### PR TITLE
Memberships: prevent overwriting product title when default name is used

### PR DESCRIPTION
**Pull request by chrismccluskey**
Related to 248-gh-automattic/gold

## Proposed Changes

* Modify detection of default name for product title so that default names can be used so long as it's not the default for the selected schedule.

## Testing Instructions

* Use the calypso live links and navigate to the monetize features (tools > monetize) for a site with paid memberships plans.
* Open the create plan dialog and change the schedule from monthly to yearly, one-time and then back to monthly. Observe with each change how the title changes.
* Create a monthly plan named "Subscription" and save that plan.
* Edit the plan, and observe that the title is still named "Subscription".
* Change the schedule to "one-time" and then to "monthly" and then to "yearly". Observe how the title does not change.
* Close the edit plan dialog and click create new plan.
* Change the name of the plan. Now try changing the schedule - the title should not change.
* Change the name of the plan to "Subscription", then change the schedule to "one-time" and back to "monthly" or "yearly" - the title should not change.

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/earn/components/add-edit-plan-modal/index.tsx [hunk 1]
```diff
@@ -139,6 +139,8 @@ const RecurringPaymentsPlanAddEditModal = ( {
 	const [ editedSchedule, setEditedSchedule ] = useState(
 		product?.renewal_schedule ?? PLAN_MONTHLY_FREQUENCY
 	);
+	const previousScheduleRef = useRef( product?.renewal_schedule ?? PLAN_MONTHLY_FREQUENCY );
```
Comment (millerf) [inline client/my-sites/earn/components/add-edit-plan-modal/index.tsx]: Nice use of `useRef`

Comment (millerf) [reply client/my-sites/earn/components/add-edit-plan-modal/index.tsx]: I would have written
```
const previousScheduleRef = useRef( editedSchedule );
```
as it is defined right above...

Comment (millerf) [reply client/my-sites/earn/components/add-edit-plan-modal/index.tsx]: See https://github.com/Automattic/gold/issues/248#issuecomment-1857670698, but I ultimately believe basing everything on the plan frequency is a bit "weak".
I'd rather use if the product id is set AND add a state for [the "dirtiness" of the name input](https://www.w3.org/TR/2011/WD-html5-20110525/the-input-element.html#concept-input-value-dirty-flag).

Comment (chrismccluskey) [reply client/my-sites/earn/components/add-edit-plan-modal/index.tsx]: > I would have written

Yes, that would have been better. Not sure why I didn't think of it then.

> I'd rather use if the product id is set AND add a state for the "dirtiness" of the name input.

I considered using a dirty bit as well but didn't want to track another variable for this one purpose. The refactor I had at this commit asked the question "did the title match the default title for the schedule _prior_ to the user changing the schedule?" If yes, it considered a the title unchanged, as it was identical to the default title for that schedule.

I previously thought we wanted to maintain the changing of the title for edited products as well, but after discussing this with you on our 1:1 I understand that it's perhaps better to not change the title of edited products. I was doing this with `product.title` but I can use `product.id` instead.

I'll try the `nameWasEdited` / `nameWasTouched` approach next and update the test plan.

**Review by millerf — COMMENTED** [review]
LGTM and works as expected

**Comment by millerf** [comment]
https://github.com/Automattic/wp-calypso/assets/790558/af204ae7-2533-4054-aaf6-228503c63f31

**Review by millerf — APPROVED** [review]

**Pull request merged**